### PR TITLE
Date headers

### PR DESCRIPTION
**Pull request by alphapapa**
This seems to be working correctly, but I'm not very happy with the code, because it seems very complicated.  Maybe it's a consequence of using an Emacs buffer to store and manipulate structured, ordered data, or maybe I'm not a very good programmer.  If you can simplify the code, you're welcome to do so.  :)

I'd also appreciate if you can do some quick testing, to make sure that messages are inserted under the right headers, in the right order, etc.  I think it's finally working right, both when inserting new messages and old ones, but I wouldn't be surprised if there are a few bugs still.  

And if this is more than you feel like reviewing anytime soon, that's fine too.  I'll keep using it myself and fix any bugs I find.

Thanks.

### CODE REVIEW

**Comment by alphapapa** [comment]
Note to me/us: In the current state of this branch, when a room's history is first loaded, the message de-duplication doesn't work.  :(

**Comment by alphapapa** [comment]
Okay, I think it's working correctly now!  The code is a bit easier to follow, as well.

File: matrix-client.el [hunk 1]
```diff
@@ -493,32 +507,73 @@ If MESSAGE is nil, clear existing message."
              (handler (a-get (oref con :event-handlers) type)))
     (funcall handler con room item)))
 
+(defun matrix-client--get-date-header (timestamp)
+  "Return position of appropriate date header in current buffer for TIMESTAMP.
+Creates a new header if necessary."
+  (cl-labels ((prev-header-pos () (matrix--prev-property-change (point) 'matrix-header-day-number))
```
Comment (jgkamat) [inline matrix-client.el]: Oh, I've always wanted this and I never knew how to do it cleanly, I'll have to remember this :)

Comment (alphapapa) [reply matrix-client.el]: Yeah, I don't see it used very often in elisp, but I think it can clarify and simplify code a lot.

**Comment by alphapapa** [comment]
> Once ng-client is in, we could revisit this to make it a lot cleaner :)

Yeah, I was thinking a few minutes ago, maybe we should factor out the buffer-related code into a library...  We could call it something like `mui`...  On the other hand, I hate to reimplement the wheel since there's `lui`, but it would give us more freedom rather than depending on them to add features we need.  Something to think about anyway.

**Pull request merged**